flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/DefaultStreamGraphHasher.java

```diff
@@ -0,0 +1,313 @@
+			StreamNode node,
+			Hasher hasher,
+			Map<Integer, byte[]> hashes,
+			boolean isChainingEnabled) {
+
+		// Include stream node to hash. We use the current size of the computed
+		// hashes as the ID. We cannot use the node's ID, because it is
+		// assigned from a static counter. This will result in two identical
+		// programs having different hashes.
+		generateNodeLocalHash(node, hasher, hashes.size());
+
+		// Include chained nodes to hash
+		for (StreamEdge outEdge : node.getOutEdges()) {
+			if (isChainable(outEdge, isChainingEnabled)) {
+				StreamNode chainedNode = outEdge.getTargetVertex();
+
+				// Use the hash size again, because the nodes are chained to
+				// this node. This does not add a hash for the chained nodes.
+				generateNodeLocalHash(chainedNode, hasher, hashes.size());
+			}
+		}
+
+		byte[] hash = hasher.hash().asBytes();
+
+		// Make sure that all input nodes have their hash set before entering
+		// this loop (calling this method).
+		for (StreamEdge inEdge : node.getInEdges()) {
+			byte[] otherHash = hashes.get(inEdge.getSourceId());
+
+			// Sanity check
+			if (otherHash == null) {
+				throw new IllegalStateException("Missing hash for input node "
+						+ inEdge.getSourceVertex() + ". Cannot generate hash for "
+						+ node + ".");
+			}
+
+			for (int j = 0; j < hash.length; j++) {
+				hash[j] = (byte) (hash[j] * 37 ^ otherHash[j]);
+			}
+		}
+
+		if (LOG.isDebugEnabled()) {
+			String udfClassName = "";
+			if (node.getOperator() instanceof AbstractUdfStreamOperator) {
+				udfClassName = ((AbstractUdfStreamOperator<?, ?>) node.getOperator())
+						.getUserFunction().getClass().getName();
+			}
+
+			LOG.debug("Generated hash '" + byteToHexString(hash) + "' for node " +
+					"'" + node.toString() + "' {id: " + node.getId() + ", " +
+					"parallelism: " + node.getParallelism() + ", " +
+					"user function: " + udfClassName + "}");
+		}
+
+		return hash;
+	}
+
+	/**
+	 * Applies the {@link Hasher} to the {@link StreamNode} (only node local
+	 * attributes are taken into account). The hasher encapsulates the current
+	 * state of the hash.
+	 * <p>
+	 * <p>The specified ID is local to this node. We cannot use the
+	 * {@link StreamNode#id}, because it is incremented in a static counter.
+	 * Therefore, the IDs for identical jobs will otherwise be different.
+	 */
+	protected void generateNodeLocalHash(StreamNode node, Hasher hasher, int id) {
+		// This resolves conflicts for otherwise identical source nodes. BUT
+		// the generated hash codes depend on the ordering of the nodes in the
+		// stream graph.
+		hasher.putInt(id);
+
+		if (node.getOperator() instanceof AbstractUdfStreamOperator) {
+			String udfClassName = ((AbstractUdfStreamOperator<?, ?>) node.getOperator())
+					.getUserFunction().getClass().getName();
+
+			hasher.putString(udfClassName, Charset.forName("UTF-8"));
+		}
+	}
+
+	private boolean isChainable(StreamEdge edge, boolean isChainingEnabled) {
```

Comment:
On a very general side, I actually think the whole hashing mechanics to assign state is very fragile and not well maintainable w.r.t. changes. In the middle to long run, we should think about an alternative. In particular when we want to remove the current chaining, backwards compatibility will become a nightmare.

For the time being, the current default will/might also become a legacy at some point in time. I wonder if it makes sense to call this thing "default" after all. Maybe those things should just be versioned and the latest version is implicitly default. Actually I was wondering if the "right" approach is to always copy the code for each versions. Inheriting from some default will break the legacy if somebody touches default before making the current default a legacy. Tricky.